### ISSUE TITLE
auto-pairs doesn't detect imbalanced pairs

### ISSUE THREAD

**Issue by MisterTicot** [jiangmiao/auto-pairs#270]
In a situation like this:

```
if (x) {
  if (y) {
    |
}
```
Pressing `}` will jump to the last bracket rather than inserting one. I think it should detect that pair are imbalanced and insert a closing bracket instead.

**Comment by nkakouros** [comment]
Another example that I find myself quite often in is this:

```bash
echo "$|array[@]}"  # need to add `{`
```

or 

```bash
echo |$[array[@]}"  # need to add `"`
```

**Comment by mversic** [comment]
yes, in it's current state this feature is unusable because I almost always want to add closing brace, because it's missing or otherwise I wouldn't be pressing it. I found that you can use `let g:AutoPairsMultilineClose = 0` to disable this, but then you loose this feature entirely, even when it makes sense. It's a trade off. I find it better disabled

If AutoPairs plugin did some scope analysis it could support this feature in a sane way, otherwise it's just silly how it behaves. I haven't found other plugin that does this well.
